### PR TITLE
implemented new mixer api1

### PR DESCRIPTION
**Pull request by AhmedKorim**
Gonna test this with dApp locally

### CODE REVIEW

**Comment by drewstone** [comment]
Whats the status of this?

**Comment by AhmedKorim** [comment]
Still need a test with the dApp, it should though

**Comment by drewstone** [comment]
Ok, the build doesn't work and would be worth fixing as well.

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 1]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
-      });
+  /**
+   *
+   *  preGenerateBulletproofGens
+   *  @description generates Bulletproof should be cached to faster future init
+   * */
+
+  public async preGenerateBulletproofGens(): Promise<void> {
+    try {
+      const wasm = await WasmMixer.wasm;
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:73]: Can we add types to this?

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 2]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
+      const opts = new wasm.PoseidonHasherOptions();
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:74]: Can we add types to this?

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 3]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
-      });
+  /**
+   *
+   *  setBulletProofGens
+   *  @description Setts the PoseidonHasher on the wasm mixer for future usage
+   *  this should be called one time
+   * */
+
+  public setBulletProofGens(bulletProofGens: Uint8Array) {
+    import('@webb-tools/wasm-utils').then((wasm) => {
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:92]: Can we use the streaming API?

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 4]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
-        return this.emit('generateNoteAndLeaf', {
-          leaf,
-          note
-        });
-      }
-    } catch (e) {
-      this.emit('generateNoteAndLeaf', e, true);
-    }
+  private static get wasm() {
+    return import('@webb-tools/wasm-utils');
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:131]: Can we use the streaming API for this?

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 5]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
-      return;
-    }
+  /**
+   *  createProof
+   *  @description Generates a note from , this will create two random numbers R wish is the secret and Nullifer both are random
+   *  @param {string} note - Note serialized
+   *  @Param {Uint8Array}  root - Merkle root for verifying against
+   *  @Param {Uint8Array[]} leaves - Adding the leaves in the merkle tree, for generating intermediate hashes
+   * */
+  public async createProof(note: string, root: Uint8Array, leaves: Array<Uint8Array>): Promise<void> {
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:140]: Lint fix needed.

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 6]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
-    leaves: Array<Uint8Array>,
-    bulletproofGens?: Uint8Array
-  ): void {
-    if (!this.mixer) {
-      this.emit('createProof', 'Mixer is not initialized', true);
-      return;
-    }
+  /**
+   *  createProof
+   *  @description Generates a note from , this will create two random numbers R wish is the secret and Nullifer both are random
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:135]: This is wrong. This seems to be a comment for `generateNote` not `createProof`.

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 7]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
-          this.logger.error(`Failed to initialize the mixer for createProofer`, e);
-          this.emit('createProof', e, true);
-        });
+      const hasher = this.hasher;
+      const wasm = await WasmMixer.wasm;
+      const myNote = wasm.Note.deserialize(note);
+      const noteGenerate = new wasm.NoteGenerator(hasher);
+      const leaf = noteGenerate.leaf_of(myNote);
+      const merkleTree = new wasm.MerkleTree(32, hasher);
+      merkleTree.push_leaf(leaf);
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:148]: What is going on here? I don't think this is correct.

File: packages/sdk-mixer/src/wasm-thread.ts [hunk 8]
```diff
@@ -64,167 +53,110 @@ export type EventRX<T extends keyof WasmWorkerMessageRX = any> = {
...
+      merkleTree.add_leaves(leaves, root);
+      const { comms, nullifier_hash, leaf_index_comms, proof, proof_comms } = merkleTree.create_zk_proof(root, myNote);
```
Comment (drewstone) [inline packages/sdk-mixer/src/wasm-thread.ts:150]: Need to fix lint.

**Review by drewstone — CHANGES_REQUESTED** [review]
Some things look good but others are still off. I think we might be handling createProof incorrectly but will need you to speak more to what _you_ think is going on.

**Comment by shekohex** [comment]
I guess we changed the build & build-release scripts, but we missed the lint one 😅

**Comment by AhmedKorim** [comment]
We need to have Typescript pushed without building that is strange to have but let's see what we can do about it

**Pull request merged**